### PR TITLE
internal/database: Do not wait for rate limiter inside transaction

### PR DESCRIPTION
**Pull request by indradhanush**
Load up all the repo names in memory and sync them one at a time instead of holding onto the transaction waiting for the rate limiter.

Earlier today, we had an incident where the repo table was locked from all queries. As a side effect of that investigation, we also found out that this [query was running](https://sourcegraph.slack.com/archives/C04SLUZNTLM/p1678178077061789) for ~1.5 hours:

```
-- query hash: 804980436
-- query length: 281 (0 args)
-- caller: internal/database.(*gitserverRepoStore).IterateWithNonemptyLastError
-- source: internal/database/gitserver_repos.go:152

SELECT
	repo.name
FROM repo
JOIN gitserver_repos gr ON repo.id = gr.repo_id
JOIN external_service_repos esr ON repo.id = esr.repo_id
JOIN external_services es on esr.external_service_id = es.id
WHERE
	gr.last_error != ''
	AND repo.deleted_at IS NULL
	AND es.cloud_default IS TRUE
```

While this was not the source of the problem - rather a side effect of the deadlock on the table, we can refactor this to load up the repo names in memory instead of scanning through one at a time and waiting on the rate limiter. In a worst case scenario where the syncer is rate limited, this will unnecessarily wait for a long enough time until the rate limiter is available or the PG backend cancels the query.

## Test plan

- Updated existing tests
- Build should pass

### CODE REVIEW

File: internal/database/gitserver_repos.go [hunk 1]
```diff
@@ -148,28 +148,19 @@ WHERE
...
-		if err := rows.Scan(&name); err != nil {
-			return errors.Wrap(err, "scanning row")
-		}
-		err := repoFn(name)
-		if err != nil {
-			// Abort
-			return errors.Wrap(err, "calling repoFn")
-		}
-	}
+type lastErroredRepo struct {
```
Comment (sashaostrikov) [inline internal/database/gitserver_repos.go:156]: looks like a leftover code for me

File: internal/database/gitserver_repos.go [hunk 2]
```diff
@@ -185,6 +176,8 @@ WHERE
 	AND es.cloud_default IS TRUE
 `
 
+var scanLastErroredRepos = basestore.NewSliceScanner(scanLastErroredRepoRow)
```
Comment (sashaostrikov) [inline internal/database/gitserver_repos.go:179]: nit: I'd move it closer to `scanLastErroredRepoRow`

File: internal/database/gitserver_repos.go [hunk 3]
```diff
@@ -148,28 +148,19 @@ WHERE
 	AND es.cloud_default IS TRUE
 `
 
-func (s *gitserverRepoStore) IterateWithNonemptyLastError(ctx context.Context, repoFn func(repo api.RepoName) error) (err error) {
+func (s *gitserverRepoStore) ListReposWithLastError(ctx context.Context) ([]*api.RepoName, error) {
```
Comment (sashaostrikov) [inline internal/database/gitserver_repos.go:151]: I cannot tell why do we need a pointer for `api.RepoName` here. It is not used as a pointer on a call-site. Should we make it a value type instead?

**Review by sashaostrikov — APPROVED** [review]
LGTM! Added a couple of questions/suggestions.

File: internal/database/gitserver_repos.go (hunk 2 above)
Comment (indradhanush) [reply internal/database/gitserver_repos.go:179]: Agreed. I moved some code around it and left this here. 🤦🏽‍♂️

File: internal/database/gitserver_repos.go (hunk 3 above)
Comment (indradhanush) [reply internal/database/gitserver_repos.go:151]: Hmm, fair enough. 👍

**Review by mrnugget — APPROVED** [review]

**Pull request merged**